runtime/js/99_main.js

```diff
@@ -553,6 +554,43 @@ delete Intl.v8BreakIterator;
     postMessage: util.writable(postMessage),
   };
 
+  function promiseRejectCallback(type, promise, reason) {
+    switch (type) {
+      case 0: {
+        core.opSync("op_store_pending_promise_exception", promise, reason);
+        break;
+      }
+      case 1: {
+        core.opSync("op_remove_pending_promise_exception", promise);
+        break;
+      }
+      default:
+        return;
+    }
+    core.opAsync("op_next_task").then(() => {
+      const hasPendingException = core.opSync(
+        "op_has_pending_promise_exception",
+        promise,
+      );
+
+      if (!hasPendingException) {
+        return;
+      }
+
+      const event = new PromiseRejectionEvent("unhandledrejection", {
+        cancelable: true,
+        promise,
+        reason,
+      });
+      globalThis.dispatchEvent(event);
+
+      // If event was not prevented we will let Rust side handle it.
+      if (event.defaultPrevented) {
+        core.opSync("op_remove_pending_promise_exception", promise);
+      }
+    });
+  }
```

Comment:
I'm wondering if we could re-write this API completely in Rust. I'll give it a try